### DISCUSSION TITLE
Script mit Parametern

### DISCUSSION THREAD

**Discussion by hmichels** [spali/go-rscp#36] [Q&A] [answered]
Hallo Spali,

herzlichen Dank für diese Integration, ich habe lange gesucht und gebraucht um das zu Nutzen, bin jetzt aber sehr froh, dass es klappt.

Ich habe ein Script nach deinem Beispiel, dass die Ladezeiten des Akkus an allen Tagen zu bestimmten Uhrzeiten setzt. 
Dabei sind die Tage als auch die Uhrzeiten fest im "data_template" ("request") vom "service: shell_command.e3dc_request" enthalten und werden vom Script selbst als String ausgewertet.

Nun suche ich nach einer Möglichkeit die Tage oder zumindest die Start-/Stop-Uhrzeiten (Stunde, Minute) dem Script als Parameter vorzugeben. Der übliche Weg z.B. über "fields" und den Namen dann im "request" als {{ start_h }} anzugeben funktioniert leider nicht. In diesem Fall meckert HA ("...template value should be a string..."). Stellt man {{ start_h }} in Anführungszeichen (also "{{ start_h }}"), dann meckert die RSCP-Integration ("...error: message at index 0: expected *uint8 got string : value does not match data type").

Hast du eine Hilfe / Vorschlag für mich für die Lösung?

Vielen Dank und Gruß,
Holger

**Comment by spali** [comment]
Ist schwer zu sagen ohne konkretes Beispiel.
Ich denke du musst schauen das das im JSON als Zahl und nicht als Zahl in Anführungszeichen erscheint.
Könnte es sein, dass du `| tojson` vergessen hast?
Im Beispiel ist der Script aufruf:
```yaml
shell_command:
  e3dc_request: /config/packages/e3dc/e3dc_request.sh '{{ request | tojson }}'
```
request ist im Jinja oder python format von HA. `tojson` sollte dieses Objekt in ein JSON Objekt (Syntax) umwandeln.
Ich würde mal `e3dc_request.log` anschauen wie es da ankommt. Die Zahl sollte da natürlich ohne `"` drin stehen.

**Comment by hmichels** [comment]
Danke für die schnelle Rückmeldung!

Der Script-Aufruf ist genau so wie im Beispiel mit `| tojson` geschrieben.

Mein Script (Anfang) sieht so aus:
```
alias: .E3DC Ladesperre 0-11 Uhr aktivieren
fields:
  start_h:
    description: Startzeit in Stunden (h)
    example: "00"
    default: 0
  start_m:
    description: Startzeit in Minuten (m)
    example: "00"
    default: 0
  stop_h:
    description: Stopzeit in Stunden (h)
    example: "09"
    default: 9
  stop_m:
    description: Stopzeit in Minuten (m)
    example: "29"
    default: 29
sequence:
  - service: shell_command.e3dc_request
    data_template:
      request: |-
        {{ [
          [
            "EMS_REQ_SET_IDLE_PERIODS",
            [
              [
                "EMS_IDLE_PERIOD",
                [
                  ["EMS_IDLE_PERIOD_TYPE", 0],
                  ["EMS_IDLE_PERIOD_DAY", 0],
                  ["EMS_IDLE_PERIOD_ACTIVE", true],
                  [
                    "EMS_IDLE_PERIOD_START",
                    [
                      ["EMS_IDLE_PERIOD_HOUR", "{{ start_h }}"],
                      ["EMS_IDLE_PERIOD_MINUTE", "{{ start_m }}"]
                    ]
                  ],
                  [
                    "EMS_IDLE_PERIOD_END",
                    [
                      ["EMS_IDLE_PERIOD_HOUR", "{{ stop_h }}"],
                      ["EMS_IDLE_PERIOD_MINUTE", "{{ stop_m }}"]
                    ]
                  ]
                ]
              ],
```

Im Log-File e3dc_request.log kommt genau das an, was im Script steht, 
also z.B. `[["EMS_IDLE_PERIOD_HOUR", "{{ start_h }}"],`.

Schreibt man ohne `"` 
```
                      ["EMS_IDLE_PERIOD_HOUR", {{ start_h }}],
                      ["EMS_IDLE_PERIOD_MINUTE", {{ start_m }}]
```
, so meckert HA.

**Comment by spali** [comment, marked as answer]
Der Fehler liegt darin, dass der ganze Request innerhalb `{{` als Template definiert ist. Und du versuchst ein Template innert einem Template ;)
Sollte gehen:
```
                      ["EMS_IDLE_PERIOD_HOUR", stop_h],
```
Tipp: Teste den ganzen request im Entwicklerwerkzeuge unter Template. Dann siehst du was rauskommt und musst nicht "blind" arbeiten.

**Comment by hmichels** [comment]
Okay, der Vorschlag und der Hinweis mit dem Entwicklerwerkzeug hat geholfen ;-)

Das Skript selbst verwendet nun die Variablen `start_h`, `start_m`, `stop_h` und `stop_m` für alle Tage, kann so aber selbst nicht ausgeführt werden, da die sogenannten "Dienstdaten" dort nicht angegeben werden (können).

Die Lösung ist natürlich das Skript per Entwicklerwerkzeug oder schlussendlich als Automatisierung mit ggf. unterschiedlichen "Dienstdaten" auszuführen. Das hat geklappt.

Vielen herzlichen Dank für die super schnelle Hilfe!
